### PR TITLE
Null reference type annotations for the cursor API

### PR DESCRIPTION
**Pull request by Logerfo**
Work for #1107.

## Proposed changes
- Files annotated: `Cursor.cs`, `CursorConverter.cs` and `Cursors.cs`, in `/src/System.Windows.Forms/src/System/Windows/Forms/`.

## Regression? 
- No

## Risk
- If annotated wrongly, might need breaking changes to fix it.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2778)

### CODE REVIEW

**Comment by Logerfo** [comment]
@RussKie isn't the build supposed to fail here? I don't think the analyzer is enforcing NRT annotations consistence with the shipped API file...

File: src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs [hunk 1]
```diff
@@ -510,7 +508,7 @@ public override string ToString()
 
             if (!(left is null))
             {
-                return (left._handle == right._handle);
+                return (left._handle == right!._handle);
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs]: any way we can rewrite this to be clearer.

Maybe
```
if (right is null)
{
    return left is null;
}
if (left is null)
{
    return false;
}
...
}
```

Comment (Logerfo) [reply src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs]: Hopefully I didn't break the behavior. e9416d39f01385de239a3ba7ad0317fb92111b3d

File: src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs [hunk 2]
```diff
@@ -99,8 +97,8 @@ public override object ConvertTo(ITypeDescriptorContext context, CultureInfo cul
                     for (int i = 0; i < props.Length; i++)
                     {
                         PropertyInfo prop = props[i];
-                        object[] tempIndex = null;
-                        Cursor c = (Cursor)prop.GetValue(null, tempIndex);
+                        object[]? tempIndex = null;
+                        Cursor c = (Cursor)prop.GetValue(null, tempIndex)!;
                         if (c == cursor)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs]: Aren't these two lines in contradiction?

File: src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs]: To verify the correctness you can write a test before making the change and run it after the change 😉

Comment (Logerfo) [reply src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs]: I think the behavior is already being fully tested by the following test: https://github.com/dotnet/winforms/blob/fde3805ea206227ad01266c7d25ce0ed2df4996f/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CursorTests.cs#L435-L453

File: src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs [hunk 3]
```diff
@@ -65,8 +63,8 @@ public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo c
                     PropertyInfo prop = props[i];
                     if (string.Equals(prop.Name, text, StringComparison.OrdinalIgnoreCase))
                     {
-                        object[] tempIndex = null;
-                        return prop.GetValue(null, tempIndex);
+                        object[]? tempIndex = null;
+                        return prop.GetValue(null, tempIndex)!;
```
[the remark this answers is not in the document]
Comment (Logerfo) [reply src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs]: Currently, yes. Every property on `Cursors.cs` is now even NRT-annotated as non-nullable. But I couldn't find a test to ensure it for the future. Do you recommend to write it?

File: src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs (hunk 2 above)
Comment (Logerfo) [reply src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs]: I'm sorry, but I cannot see in what way they could contradict themselves. I'm also not sure if you're talking about the original lines or the new addition, but I don't see it in any way. Perhaps you misread `if (c == cursor)` as `if (c == null)`? If you still think something is wrong, please elaborate.

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs]: I think this is fine. `cursor` is a parameter to the method, we're looping through every property on `Cursors.cs` and comparing the input to a known value

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs]: I'm not sure where my eyes were yesterday. I think I was seeing `if (c == null)...`. 🤕

File: src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs]: Awesome!

**Comment by RussKie** [comment]
> isn't the build supposed to fail here? I don't think the analyzer is enforcing NRT annotations consistence with the shipped API file...

I think this may be  bug in the analyzer - it is not picking the "nullability"
* Nullable:
image
* Non nullable:
image

@mavasani any thoughts? Is this a bug or we doing something wrong?

**Comment by mavasani** [comment]
Tagging @jcouv who recently implemented the nullability annotation supported in public API analyzer. I believe you will need to pull the latest beta package from here: https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.PublicApiAnalyzers

**Comment by RussKie** [comment]
Thank you @mavasani for the heads up.

@jcouv I've pulled the latest package, and I'm getting the correct annotations generated. However I don't believe they are getting generated in the right place - I get `PublicAPI.Shipped.txt` updated instead of `PublicAPI.Unshipped.txt`. This is incorrect since we haven't shipped the NRT decorated API.

image
image

**Comment by RussKie** [comment]
I had to move all changes from `PublicAPI.Shipped.txt` to `PublicAPI.Unshipped.txt` manually (which can be labourious).

**Comment by mavasani** [comment]
Thanks. Can you file an issue here: https://github.com/dotnet/roslyn-analyzers/issues/new

**Comment by Logerfo** [comment]
@RussKie so was `TreeNodeMouseClickEventArgs.cs` the only broken file?

File: src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Logerfo) [reply src/System.Windows.Forms/src/System/Windows/Forms/CursorConverter.cs]: Here you go 5ba943f331ea5685e624a70d1bcd2383da00faef, but the CI is not running for some reason.

**Comment by RussKie** [comment]
> so was `TreeNodeMouseClickEventArgs.cs` the only broken file?

Looks like it.

**Comment by RussKie** [comment]
I've undone 0fbf166 and moved the API updates to shipped.txt file as suggested by @sharwell and @mavasani.

**Comment by RussKie** [comment]
I'm kind of speculating here, may be draft PRs don't gets built? 🤔

**Comment by Logerfo** [comment]
@RussKie yeah, I think that was it. It's failing for the newly annotated API, even though the surface is defined in the file. Any idea?

**Comment by RussKie** [comment]
I have no idea why the build fails, I am just as puzzled...
I have spent several hours and I started to think that there is something wrong with the analyzer. I have raised https://github.com/dotnet/roslyn-analyzers/issues/3260

@jcouv I would appreciate your help here.

**Comment by RussKie** [comment]
Just a heads up: we're currently blocked by our downstream consumers and unable to update to the latest .NET Runtime, which contains the latest compiler needed for the Roslyn analysers.
It maybe another month or so... I'll provide an update when the situation changes.

**Comment by RussKie** [comment]
I believe we should be unblocked now.
Can you please rebase?

**Comment by Logerfo** [comment]
It's been a while... please remind me of what are we heading towards regarding this file: 
image

**Comment by weltkante** [comment]
Other PRs are updating the `Unshipped.txt` files together with the PR, so I'd expect it to be done here as well instead of turning the entire feature off again for the project. The `Shipped.txt` files obviously only must be updated when something is shipped, so can't be part of this PR, any change there is suspicious and most likely a mistake. As far as I know the `NoWarn` isn't necessary if Unshipped files are updated correctly.

**Comment by Logerfo** [comment]
And what about the "release" condition?

**Comment by weltkante** [comment]
As far as I can tell it has been intentionally enabled for all builds, so I don't see why this PR should disable it again. You are just adding annotations for a specific API, no reason to edit any infrastructure in the same PR.

**Comment by Logerfo** [comment]
Alright then, conflicts solved. I think there's more to undo, but let's see if it builds first.

**Comment by Logerfo** [comment]
@RussKie the build still doesn't seem to recognize the nullability. Also, I'm not sure whether the current changes on the following files are required:
- SWFDesign/Shipped.txt
- SWFDesign/Unshipped.txt
- SWFPrimitives/Unshipped.txt
- SWF/Unshipped.txt

Sorry, I'm still trying to recall what exactly was going on the last time we dealt with this.

**Comment by RussKie** [comment]
> Sorry, I'm still trying to recall what exactly was going on the last time we dealt with this.

Haha, me too :)

> It's been a while... please remind me of what are we heading towards regarding this file:
> image

This is how I expect the file to look like:
```xml
<Project>

  <PropertyGroup Condition="'$(UsePublicApiAnalyzers)' == 'true'">
    <!-- https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md -->
    <WarningsAsErrors>$(WarningsAsErrors);RS0016;RS0017;RS0022;RS0024;RS0025;RS0026;RS0027;RS0036;RS0037</WarningsAsErrors>
  </PropertyGroup>

</Project>
```





> the build still doesn't seem to recognize the nullability.

I'll try to set aside some time and find out what's going on.

**Comment by RussKie** [comment]
> the build still doesn't seem to recognize the nullability.

There are few things we need to do.
1. We need to update the analyzer from 2.9.8. We can go with the latest official https://www.nuget.org/packages/Microsoft.CodeAnalysis.PublicApiAnalyzers/3.0.0-beta2.final, or with a nightly https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.PublicApiAnalyzers (need to add a new NuGet feed https://dotnet.myget.org/F/roslyn-analyzers/api/v3/index.json).
2. We need to enable nullability in the `PublicAPI.*.txt` files: https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md#nullable-reference-type-support
  NB: beware of dups https://github.com/dotnet/roslyn-analyzers/issues/3450
3. Doing \#2 will generate thousands of errors (don't ask me how I know). So we need to re-establish the baseline. This we should do as a separate PR.
4. Once \#3 is done, we'll need to rebase all NRT-related work and go from there.

Any issues with the analyzer we should raise at https://github.com/dotnet/roslyn-analyzers/issues. @mavasani @jcouv would be on a hook for these :)

/cc: @hughbe

**Comment by Logerfo** [comment]
1) It was already in a beta version, but I upgraded it as well.
2) This is already done.
3) OK, I won't ask.
4) I think it's time to hibernate again, then...

Hmmm, weird. My local build has many more build errors (from the analyzer) than the CI build.

**Comment by RussKie** [comment]
#3157 addressed point 1-3 (I believe).
Can you please rebase and see what happens?

**Comment by Logerfo** [comment]
@RussKie here you go

**Comment by RussKie** [comment]
Yep, there are still kinks to sort out, check https://github.com/dotnet/winforms/pull/2824#issuecomment-621146869 for more details

**Comment by Logerfo** [comment]
Rebased it again. I don't get it... Why does the errors come from members that were not changed by this PR?

**Comment by Logerfo** [comment]
@weltkante thanks for the clarification. It was probably something wrong with the rebases.
@RussKie the build is passing now.

File: NuGet.config [hunk 4]
```diff
@@ -7,5 +7,6 @@
     <add key="dotnet5" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json" />
     <add key="dotnet5-transport" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5-transport/nuget/v3/index.json" />
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
+    <add key="myget" value="https://dotnet.myget.org/F/roslyn-analyzers/api/v3/index.json" />
```
Comment (RussKie) [inline NuGet.config:10]: I don't think this is needed

Comment (Logerfo) [reply NuGet.config:10]:
```suggestion
```

**Comment by hughbe** [comment]
So I know you're just inheriting from `TypeConverter`. But, I noticed that `TypeConverter` is not annotated with nullable attributes yet (see 9. in https://github.com/dotnet/runtime/issues/2339)

For example, in the`GetStandardValuesSupported(ITypeDescriptorContext context)` `context` can be null, e.g.
```cs
        /// <summary>
        /// Gets a value indicating whether this object supports a standard set of values
        /// that can be picked from a list.
        /// </summary>
        public bool GetStandardValuesSupported() => GetStandardValuesSupported(null);

        /// <summary>
        /// Gets a value indicating whether this object supports a standard set of values that can be picked
        /// from a list using the specified context.
        /// </summary>
        public virtual bool GetStandardValuesSupported(ITypeDescriptorContext context) => false;
```

I don't know where you should back out the nullable changes in `TypeConverter` subclass, because as far as I am concerned we may be incorrectly annotating APIs as non-nullable when they should be

File: NuGet.config (hunk 4 above)
Comment (Logerfo) [reply NuGet.config]: Correct, it worked without it.

**Comment by Logerfo** [comment]
@hughbe I didn't know there was missing annotations in the framework, thanks for the heads up!
Possibilites are:
1) Step back from the changes in the type converter only (I don't know if it's possible, gotta read the changes again to confirm it).
2) If we can expect something, annotate the type converter considering the expected annotations coming to the framework.
3) Hold the PR until the framework gets the annotations we need (do we have an ETA?).

**Comment by RussKie** [comment]
Can we exclude `CursorConverter`?

**Comment by RussKie** [comment]
Thank you

**Pull request merged**